src/api/rpc_api/state.rs

```diff
@@ -100,6 +107,20 @@ pub trait GetStorage {
 		at_block: Option<Self::Hash>,
 	) -> Result<Vec<StorageKey>>;
 
+	/// Retrieve up to `count` keys. Support prefix and pagination support.
+	/// The number of keys returned is not limited. For big numbers, the rpc calls will be made several times.
+	/// Up to `count` keys will be returned.
+	/// If `start_key` is passed, return next keys in storage in lexicographic order.
+	///
+	/// `at_block`: the state is queried at this block, set to `None` to get the state from the latest known block.
+	async fn get_all_storage_keys_paged_up_to_count(
```

Comment:
That's how I imagined this function. But I think this version should be the default, not the original one that throws an error. 

So I'm in favour of renaming this function to `get_all_storage_keys_paged` and rename the original to something else. Not sure if we should use `get_all_storage_keys_paged_up_to_count` as a name, because that's true for both, new and original, functions.

Reply:
I think the default behaviour should be similar to that of the substrate, i.e. the number of keys returned is limited and return an error on big value of count. Users can choose to use the alternative method if they wish to bypass the limits set by substrate. 
get_all_storage_keys_paged is only returning up to count keys, if the count input value < limit. But, yes the name could be more descriptive. How about `get_all_storage_keys_paged_without_limit` or `get_all_storage_keys_paged_workaround`?

Reply:
It's just that I, while using this function, promptly run into this error quite unexpectedly and it took me some time to figure out why. I find it counterintuitive to allow a count input (as `u32` nonetheless) and limit it to `1000` during runtime (which is not documented at all: https://polkadot.js.org/docs/kusama/rpc#getkeyspagedkey-storagekey-count-u32-startkey-storagekey-at-blockhash-vecstoragekey)

So I'm not sure if we should follow the Substrate way here - I believe we can do better. Or what would be the downside of having the "workaround" as a default? Sure, it may take a long time if the user wants to fetch more than 1000 keys. In the worst case, the program even needs to be terminated. But in case an error is returned (due to count > 1000) that probably also ends in a non-functional program that needs to be terminated.

I'm not set on my idea, so if you think it's better to leave the original as default, that's fine to me.

Reply:
> So I'm not sure if we should follow the Substrate way here - I believe we can do better. Or what would be the downside of having the "workaround" as a default? 
> 
That the user is unaware of using a workaround? 

>Sure, it may take a long time if the user wants to fetch more than 1000 keys. In the worst case, the program even needs to be terminated.

If this is the biggest problem our function can cause, then yes, our workaround can be the default function.

Reply:
@Niederb  What's your opinion here?

Reply:
I would also make the new implementation the default. Also see my separate review comment

Reply:
Ok, then:
`get_storage_keys_paged ` is our workaround and  `get_storage_keys_paged_limited` is the original one, calling the RPC substrate storage_keys_paged only once.